### PR TITLE
chore: migrate to symfony/mailer

### LINKED ISSUE #40564: Swiftmailer has reached EOL in Nov '21
[according to the `core/composer.json` we are using Swiftmailer](https://github.com/owncloud/core/blob/de753b61155ff640d1bf5bff978e19233a33ab8c/composer.json#L51)

[The Swiftmailer documentation states:](https://swiftmailer.symfony.com/docs/introduction.html)
> Swift Mailer will stop being maintained at the end of November 2021.
> Swiftmailer is not maintained anymore. Use Symfony Mailer instead. Read more on Symfony's blog.

I would exepct ownCloud to use supported PHP components in case of security issues.

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Related Issue

- Fixes https://github.com/owncloud/core/issues/40564

### Scenarios
- try to send mail from admin section
- send invite via CalDAV 

### Mail Setups to test
- send mail
- qmail
- smtp with ssl
- smtp without ssl

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
@mmattel we have dropped one config setting - also the admin ui looks a bit different now .....

File: config/config.sample.php [hunk 1]
```diff
@@ -396,6 +396,7 @@
 /**
  * Enable or disable SMTP class debugging
  */
+# TODO: unused / deprecated
```
Comment (mmattel) [inline config/config.sample.php:399]: If is unused/deprecated, just remove the whole block. This was "just" a setting temporarily set to true in the past. We do not have any deprecation process except release notes AFAICT. Therefore just mention it there --> @pako81

**Comment by mmattel** [comment]
> smtp encryption option is now either 'ssl/tls' or 'none' .... 'STARTTLS' has been dropped

Same as written in the code comment above, this is a release notes thing and needs mentioning there --> @pako81

**Comment by mmattel** [comment]
> the admin ui looks a bit different now

Can one drop me a screenshot we can use for docs into: https://github.com/owncloud/docs-server/issues/1137 ? and clarify for which version this change is targeted?

@pako81 @jnweiger fyi

File: config/config.sample.php (hunk 1 above)
Comment (DeepDiver1975) [reply config/config.sample.php]: comment removed - smtp logging is back in the code base - THX fr noticing

**Comment by DeepDiver1975** [comment]
@individual-it @phil-davis can I ask you for a quick look - THX a lot:
image
image

**Comment by individual-it** [comment]
The failing case is that `sendmail` is set as the mode but no sendmail is installed/configured on the system (`/usr/sbin/sendmail -bs`).
Looks like the exception is thrown but does not cause `occ` to return `1` as exit code

File: lib/private/Mail/Mailer.php [hunk 2]
```diff
@@ -89,18 +86,33 @@ public function createMessage() {
...
+			$recipients = array_merge($message->getTo(), $message->getCc(), $message->getBcc());
+			array_walk($recipients, static function ($value, $key) use (&$failedRecipients) {
+				if (is_numeric($key)) {
+					$failedRecipients[] = $value;
+				} else {
+					$failedRecipients[] = $key;
+				}
+			});
+
+			return $failedRecipients;
```
Comment (phil-davis) [inline lib/private/Mail/Mailer.php:114]: IMO this now returns the failed recipients, but previously I think that it threw an exception.

Some callers of `mailer->send` are not checking for failed recipients, they are only catching an exception.

I adjusted some calls related to password reset emails - see #41013 - and I think that will make the tests pass.

Maybe there is a different solution that is better. Maybe more callers of `mailer->send` need to be adjusted. @DeepDiver1975 you can think about that!

Comment (DeepDiver1975) [reply lib/private/Mail/Mailer.php:114]: The API design is a bit strange.
I will look into it to see if the behavior can be adjusted to match the previous....

Comment (DeepDiver1975) [reply lib/private/Mail/Mailer.php:114]: There are a few locations (found 2) where the return value is evaluated but then just an exception is thrown.

**Review by phil-davis — COMMENTED** [review]
Looks good, and tests all pass.
I made a few notes mainly about backward-compatibility. The way the ownCloud `Mailer` etc classes are currently structured, they expose parts of SwiftMailer, so to get rid of SwiftMailer there have to be changes to public method interfaces.
What do we do about that for oC core major version 10 with PHP 7.4?

(for PHP 8.2 it is easier - we would bump the major version of oC core to 11)

File: lib/private/Mail/Message.php [hunk 3]
```diff
@@ -22,55 +22,38 @@
...
+	private array $to = [];
+	private array $cc = [];
+	private array $bcc = [];
 
-	/**
-	 * @param Swift_Message $swiftMessage
-	 */
-	public function __construct(Swift_Message $swiftMessage) {
-		$this->swiftMessage = $swiftMessage;
+	public function __construct(Email $swiftMessage) {
```
[the remark this answers is not in the document]
Comment (DeepDiver1975) [reply lib/private/Mail/Message.php:41]: Ctor is not to be called by apps. IMailer::createMessage is to be used which did not change.

File: lib/private/Mail/Message.php [hunk 4]
```diff
@@ -235,26 +197,27 @@ public function getPlainBody() {
...
 		return $this;
 	}
 
-	/**
-	 * Get's the underlying SwiftMessage
-	 * @return Swift_Message
-	 */
-	public function getSwiftMessage() {
-		return $this->swiftMessage;
+	public function getMessage(): Email {
```
[the remark this answers is not in the document]
Comment (DeepDiver1975) [reply lib/private/Mail/Message.php:205]: Due to this the original design did not have an iMessage.
Never the less: yes - this breaks API.

**Comment by phil-davis** [comment]
This can wait for merge until release-10.13.2 is out of the way, so we have time to discuss what happens for releases after doing this change for the PHP 7.4 code base - will it be "the usual" 10.14.0 minor version bump, and we work out how to address backward-compatibility issues (for example, that any real backward-compatibility issues can be handled OK for any apps that are affected).

Or would this cause a major version bump, forcing an 11.0.0 release that supports PHP 7.4...

**Comment by DeepDiver1975** [comment]
From my understanding oc11 is php8.2
We can also merge this or into the php8.2 pr

**Comment by phil-davis** [comment]
I suppose this is blocked/stalled for now. Can't easily be part of 10.14 because of backward-compatibility issues that are not easily avoided.

**Comment by DeepDiver1975** [comment]
I planned to do this as part of #40981 :shrug:

**Comment by DeepDiver1975** [comment]
cleaning up - I think this is no longer of anybody's interest

**Pull request closed without merging**